presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregationFunction.java

```diff
@@ -0,0 +1,338 @@
+    {
+        super(name,
+                ImmutableList.of(orderableTypeParameter("E")),
+                ImmutableList.of(),
+                parseTypeSignature("E"),
+                ImmutableList.of(parseTypeSignature("E")));
+        requireNonNull(operatorType);
+        this.operatorType = operatorType;
+    }
+
+    @Override
+    public InternalAggregationFunction specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = boundVariables.getTypeVariable("E");
+        MethodHandle compareMethodHandle = functionRegistry.getScalarFunctionImplementation(internalOperator(operatorType, BOOLEAN, ImmutableList.of(type, type))).getMethodHandle();
+        return generateAggregation(type, compareMethodHandle);
+    }
+
+    protected InternalAggregationFunction generateAggregation(Type type, MethodHandle compareMethodHandle)
+    {
+        DynamicClassLoader classLoader = new DynamicClassLoader(AbstractMinMaxAggregationFunction.class.getClassLoader());
+
+        List<Type> inputTypes = ImmutableList.of(type);
+
+        MethodHandle inputFunction;
+        MethodHandle combineFunction;
+        MethodHandle outputFunction;
+        Class<? extends AccumulatorState> stateInterface;
+        AccumulatorStateSerializer<?> stateSerializer;
+
+        if (type.getJavaType() == long.class) {
+            stateInterface = NullableLongState.class;
+            stateSerializer = StateCompiler.generateStateSerializer(stateInterface, classLoader);
+            inputFunction = LONG_INPUT_FUNCTION;
+            combineFunction = LONG_COMBINE_FUNCTION;
+            outputFunction = LONG_OUTPUT_FUNCTION;
+        }
+        else if (type.getJavaType() == double.class) {
+            stateInterface = NullableDoubleState.class;
+            stateSerializer = StateCompiler.generateStateSerializer(stateInterface, classLoader);
+            inputFunction = DOUBLE_INPUT_FUNCTION;
+            combineFunction = DOUBLE_COMBINE_FUNCTION;
+            outputFunction = DOUBLE_OUTPUT_FUNCTION;
+        }
+        else if (type.getJavaType() == Slice.class) {
+            stateInterface = SliceState.class;
+            stateSerializer = StateCompiler.generateStateSerializer(stateInterface, classLoader);
+            inputFunction = SLICE_INPUT_FUNCTION;
+            combineFunction = SLICE_COMBINE_FUNCTION;
+            outputFunction = SLICE_OUTPUT_FUNCTION;
+        }
+        else if (type.getJavaType() == boolean.class) {
+            stateInterface = NullableBooleanState.class;
+            stateSerializer = StateCompiler.generateStateSerializer(stateInterface, classLoader);
+            inputFunction = BOOLEAN_INPUT_FUNCTION;
+            combineFunction = BOOLEAN_COMBINE_FUNCTION;
+            outputFunction = BOOLEAN_OUTPUT_FUNCTION;
+        }
+        else {
+            stateInterface = BlockState.class;
+            stateSerializer = new BlockStateSerializer(type);
```

Comment:
I think it would be much cleaner to squash these two commits; I was very confused when I first saw this given this is not an optimized serializer.